BrainPortal/lib/request_helpers.rb

```diff
@@ -66,5 +67,82 @@ def ban_ip(message)
     false
   end
 
+  module ClassMethods
+
+    # Returns the list registered allowed params
+    # for some actions filtered by the controller helper
+    # method spurious_params_ban_ip
+    def allowed_params_by_action #:nodoc:
+      @allowed_params_by_action || {}
+    end
+
+    # This is a helper method for controllers.
+    # Register for an action the list of param keys
+    # that are allowed; if any extra keys are found in
+    # a request, the IP address is banned using ban_ip.
+    # For example:
+    #
+    #   (In SessionsController)
+    #   spurious_params_ban_ip :create => [ :login, :password ]
+    #
+    # You can provide first a plain list of actions names,
+    #  which imply "actionname" => [], followed by a hash with the
+    # action names as keys and the allowed params as values:
+    #
+    #   spurious_params_ban_ip :new, :destroy, :show,
+    #      :create  => [ :login, :pw ],
+    #      :inspect => [ :verbose ]
+    #
+    # is the same as
+    #
+    #   spurious_params_ban_ip :new => [], :destroy => [], :show => [],
+    #      :create  => [ :login, :pw ],
+    #      :inspect => [ :verbose ]
+    #
+    def spurious_params_ban_ip(*action_params)
+      @allowed_params_by_action ||= {}
+      action_params.each do |item|
+        if item.is_a?(Hash)
+          item.each { |action,ok_params| @allowed_params_by_action[action.to_s] = Array(ok_params) }
+        else
+          @allowed_params_by_action[item.to_s] = []
+        end
+      end
+      before_action :ban_ip_on_spurious_params, :only => @allowed_params_by_action.keys
+    end
+
+  end
+
+  STANDARD_PARAMS_THAT_DO_NOT_TRIGGER_BANS = %w( id utf8 commit authenticity_token controller action cbrain_api_token format )
+  MINIMUM_NUMBER_OF_SPURIOUS_PARAMS = 3 # we need at least that many spurious params to trigger a ban
+
+  # Called as a before_action method for actions that
+  # need to have a specific list of params.
+  def ban_ip_on_spurious_params #:nodoc:
+
+    # Does the current action have a list of allowed params?
+    ok_params     = self.class.allowed_params_by_action[params[:action].to_s]
+    return true if ok_params.nil? # no list for this action, so make no further verifs
+
+    # Find out the ones we don't expect
+    extended_keys = ok_params.map(&:to_s) + STANDARD_PARAMS_THAT_DO_NOT_TRIGGER_BANS
+    spurious_keys = params.keys.map(&:to_s) - extended_keys
+    return true if spurious_keys.empty? # everything is good
+
+    # Prepare message
+    num_spurious = spurious_keys.size
+    first_four   = spurious_keys[0..3].join(", ")
```

Comment:
Why you do not use `MINIMUM_NUMBER_OF_SPURIOUS_PARAMS` here ?

Reply:
Because it's not the same. The constant is the threshold that triggers bans or not. The 0..3 here is just what we report, which is always the first four params no matter what the threshold is.